---
title: "SnRNA-Seq_Analysis"
author: "Hugues Escoffier"
date: "2023-02-04"
output: html_document
---

An automated analysis for SnRNA-Seq mouse data.
R Version 4.2.2 

# Librairies 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Install libraries
library(Seurat)
library(Signac)
library(clustree)
# Additional
library(patchwork)
library(dplyr)
library(ggplot2)
library(here)
library(MASS)
library(future)
```

# Load functions 

```{r}
# From path to SeuratObject  
path_to_object <- function(list_of_path, list_of_name, project_name = 'Default', data_type = 'Multi') {
  all_seurat_object = list()
  for (i in 1:length(list_of_path)) {
    x.data <- Read10X(data.dir = list_of_path[[i]])
    # Initialize the Seurat object with non-normalized data.
    if (data_type == 'Multi') {
      x.rna <- x.data$`Gene Expression`
    }
    # Create a Seurat Object + Add the mitochondrial information
    x.rna <- CreateSeuratObject(counts = x.rna, project = list_of_name[[i]], min.cells = 3, min.features = 200)
    x.rna[["percent.mt"]] <- PercentageFeatureSet(x.rna, pattern = "^mt-")
    all_seurat_object <- append(all_seurat_object, x.rna)
  }
  # Merging time 
  if (length(all_seurat_object) > 1) {
    combined <- merge(x = all_seurat_object[[1]], y = all_seurat_object[c(-1)], add.cell.ids = list_of_name, project = project_name)
    # Plot the number of nuclei
    print(table(combined$orig.ident))
    is.alone <- FALSE
    return(c(combined, is.alone)) 
  }
  else {
    input_data.list <- SplitObject(object = all_seurat_object[[1]], split.by = "orig.ident")
    is.alone <- TRUE
    return(c(all_seurat_object[[1]], is.alone))
  }
}

# Normalization, scaling, PCA & Dim
snrna_seq_analysis_1 <- function(input_data_w_nb, lim_nFeature_RNA_sup = 200, lim_nFeature_RNA_inf = 2500, th_percent.mt = 0.8, normalisation = 'Log'){
  is.alone <- input_data_w_nb[[2]]
  input_data <- input_data_w_nb[[1]]
  input_data.list <- SplitObject(object = input_data, split.by = "orig.ident")
  for (i in 1:length(x = input_data.list)) {
    # Visualization before trimming
    plot_trimming(input_data.list[[i]])
    # Trimming
    input_data.list[[i]] <- subset(input_data.list[[i]], subset = nFeature_RNA > lim_nFeature_RNA_sup & nFeature_RNA < lim_nFeature_RNA_inf & percent.mt < th_percent.mt)
    # Visualization after trimming
    plot_trimming(input_data.list[[i]])
  }
  if (normalisation == 'Log') {
    # Normalization
    for (i in 1:length(x = input_data.list)) {
      input_data.list[[i]] <- NormalizeData(input_data.list[[i]], normalization.method = "LogNormalize", scale.factor = 10000)
      # Calculate a subset of features that exhibit high cell-to-cell variation in the data set
      input_data.list[[i]] <- FindVariableFeatures(input_data.list[[i]], selection.method = "vst", nfeatures = 2000)
    }
  }
  else {
    # Normalization with SCTransform 
    if (is.alone == TRUE) {
      input_data <- SCTransform(input_data, vst.flavor = "v2", verbose = FALSE, vars.to.regress = "percent.mt")
    }
  }
  # Integration 
  if (is.alone == FALSE) {
    if (normalisation == 'Log') {
      combined.anchors <- FindIntegrationAnchors(object.list = input_data.list, dims = 1:30)
      input_data <- IntegrateData(anchorset = combined.anchors, dims = 1:30)
    }
    else {
      input_data.list <- lapply(X = input_data.list, FUN = SCTransform)
      features <- SelectIntegrationFeatures(object.list = input_data.list, nfeatures = 5000)
      input_data.list <- PrepSCTIntegration(object.list = input_data.list, anchor.features = features)
      combined.anchors <- FindIntegrationAnchors(object.list = input_data.list, normalization.method = 'SCT', anchor.features = features)
      input_data <- IntegrateData(anchorset = combined.anchors, normalization.method = 'SCT')
    }
    DefaultAssay(input_data) <- "integrated"
  }
  else {
    input_data <- input_data.list[[i]]
  }
  # Scaling
  all.genes <- rownames(input_data)
  input_data <- ScaleData(input_data, features = all.genes)
  # PCA + Plot
  input_data <- RunPCA(input_data, npcs=50, features = VariableFeatures(object = input_data), verbose = FALSE)
  print(input_data[["pca"]], dims = 1:5, nfeatures = 5)
  print(DimPlot(input_data, reduction = "pca"))
  # Elbow Plot
  print(ElbowPlot(input_data, ndims = 50))
  return(c(input_data, is.alone))
}

# UMAP, Markers & Heat map 
snrna_seq_analysis_2 <- function(input_data_w_nb, dimensionality, choose.resolution, hm_gene_nb = 5) {
  is.alone <- input_data_w_nb[[2]]
  input_data <- input_data_w_nb[[1]]
  # Clustering
  input_data <- FindNeighbors(input_data, reduction = "pca", dims = 1:dimensionality, verbose = FALSE)
  resolution.range <- seq(from = 0, to = 1.2, by = 0.2)
  input_data.clustree <- FindClusters(object = input_data, resolution = resolution.range)
  input_data <- FindClusters(object = input_data, resolution = choose.resolution)
  input_data <- RunUMAP(input_data, dims = 1:dimensionality)
  DefaultAssay(input_data) <- 'RNA'
  #Plot
  print(DimPlot(input_data, reduction = "umap", label = TRUE))
  if (is.alone == FALSE) {
    print(DimPlot(object = input_data, reduction = "umap", group.by = "orig.ident", label = FALSE))
    print(DimPlot(input_data, reduction = "umap", split.by = "orig.ident", label = TRUE))
    DefaultAssay(input_data) <- "integrated"
  }
  input_data.markers <- FindAllMarkers(input_data, min.pct = 0.25, logfc.threshold = 0.25, only.pos = TRUE)
  print(input_data.markers %>% group_by(cluster) %>% slice_max(n = 5, order_by = avg_log2FC))
  input_data.markers %>% group_by(cluster) %>% top_n(n = hm_gene_nb, wt = avg_log2FC) -> topX
  print(DoHeatmap(input_data, features = topX$gene) + NoLegend())
  print(head(input_data@meta.data))
  # print(clustree(input_data.clustree, prefix = 'RNA_snn_res.'))
  return(input_data)
}

# Visualization of the QC 
plot_trimming <- function(x) {
  # Plots
  VnPlot <- VlnPlot(x, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
  RNAcount_MTPer <- FeatureScatter(x, feature1 = "nCount_RNA", feature2 = "percent.mt")
  RNAcount_RNAFeat <- FeatureScatter(x, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
  # Visualize
  print(VnPlot)
  print(RNAcount_MTPer + RNAcount_RNAFeat)
}
```

# Analyse snRNA-Seq data

```{r}
example <- path_to_object(c('data/Ctl-D/outs/filtered_feature_bc_matrix', 'data/Denerve/outs/filtered_feature_bc_matrix'), c('Control', 'Denerve')) 
example2 <- snrna_seq_analysis_1(example)
```

```{r}
example <- snrna_seq_analysis_2(example2, 13, 0.4)
```

# Subset

```{r}
list_of_cluster <- c(0,1,2,3,4)
# Create a subset 
new_subset <- subset(x = input_data, idents = list_of_cluster, invert = TRUE)
# Analyse your subset 
new_subset <- snrna_seq_analysis_2(c(new_subset, 'simple'), 16, 0.4)
```

# Feature Plot 

```{r}
FeaturePlot(example, features = c("Myh1"))
```

