---
title: "Ultimate_SnRNA"
author: "Hugues Escoffier"
output: html_document
---

# SnRNA-seq 

The sequencing data preprocessing steps, including base calling, mapping and read counting, have been done using Cell Ranger. 
Based on Zhisong He & Barbara Treutlein Analysis (https://github.com/quadbiolab)

## 0. Libraries 

```{r}
# library(destiny)
# CRAN 
library(Seurat)
library(harmony)
library(dplyr)
library(patchwork)
library(stats)
library(ggplot2)
library(ggrepel)
# BioConductor 
library(destiny)
# Devtools 
library(presto)
# Genomes
```

## 1. Create Seurat Object 

```{r}
counts <- Read10X(data.dir = "data/KO/outs/filtered_feature_bc_matrix")
seurat.rna <- CreateSeuratObject(counts$`Gene Expression`, project="KO")
```

## 2. Quality Control 
We filter : 
* Cells with too few genes detected. _(representing cells which are not sequenced deep enough)_
* Cells with too many genes detected. _(corresponding to multiplets)_
* Cells with high mitochondrial transcript percentage.

```{r}
# Add mitochondrial transcript percentage
seurat.rna[["percent.mt"]] <- PercentageFeatureSet(seurat.rna, pattern = "^mt-")
# Visualize 
VlnPlot(seurat.rna, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=0)
FeatureScatter(seurat.rna, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(seurat.rna, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
# The number of detected genes and number of detected transcripts are well correlated across cells while mitochondrial transcript percentage is not.
```

```{r}
# Same thresholds as Lin et al. 
seurat.rna <- subset(seurat.rna, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 0.8)
```

## 3. Normalization 
Since the amount of captured RNA can vary between cells, it is not appropriate to compare the number of captured transcripts for individual genes directly. To ensure that gene expression levels can be accurately compared across different cells, it is necessary to perform a normalization step. This step is aimed at making gene expression levels between cells comparable.

```{r}
seurat.rna <- NormalizeData(seurat.rna)
```

## 4. Feature selection
Identification of highly variable features/genes, which are genes with the most varied expression levels across cells.

```{r}
# There are no defined values in the literature. Generally we choose values between 2000 and 5000. The results do not differ significantly between these values. 
seurat.rna <- FindVariableFeatures(seurat.rna, nfeatures = 3000)
```

```{r}
# Visualize
top_features <- head(VariableFeatures(seurat.rna), 20)
plot1 <- VariableFeaturePlot(seurat.rna)
plot1
LabelPoints(plot = plot1, points = top_features, repel = TRUE)
```

## 5. Scaling 
If no data transformation is performed, the analysis could be biased towards highly expressed genes since different genes have varying base expression levels and distributions. This is an undesirable outcome as we want our analysis to consider all genes equally. To address this issue, we apply scaling to the data using selected features, which is a common practice in various data science fields. This ensures that the contribution of each gene to the analysis is appropriately accounted for, regardless of their expression levels.

```{r}
seurat.rna <- ScaleData(seurat.rna, vars.to.regress = c("nFeature_RNA", "percent.mt")) # Remove unwanted sources of variation
```

Please note that steps 3 to 5 can be replaced by a `SCTransform` step

## 6. PCA

```{r}
seurat.rna <- RunPCA(seurat.rna, npcs = 50)
ElbowPlot(seurat.rna, ndims = ncol(Embeddings(seurat.rna, "pca")))
# One would assume that the first phase of the curve represents the 'real' signal related to biological differences between cell populations, while the second phase mostly represents technical variation or the stochastic nature of individual cells.
# * It is very difficult to precisely define the elbow point or turning point of the curve, as it is usually not a perfect elbow.
# * Higher-ranked PCs do explain more variation than lower-ranked PCs, but more explained variations does not necessarily mean higher information content. Sometimes there is interesting but weak signal buried in the noise and therefore as part of lower-ranked PCs.
```

## 7. Non-linear dimension reduction 

TSNE provides great visualization when cells form distinct cell groups, while UMAP perserves trajectory-like structure better when data contains 'continuum'.
```{r}
seurat.rna <- RunTSNE(seurat.rna, dims = 1:15)
seurat.rna <- RunUMAP(seurat.rna, dims = 1:15)
plot1 <- TSNEPlot(seurat.rna)
plot2 <- UMAPPlot(seurat.rna)
plot1
plot2
```

```{r}
FeaturePlot(seurat.rna, c("Myh1","Myh2","Myh4","Myh7","Xist","Uty","Col22a1","Pam","Ttn"),
ncol=3, reduction = "tsne")
FeaturePlot(seurat.rna, c("Myh1","Myh2","Myh4","Myh7","Xist","Uty","Col22a1","Pam","Ttn"),
ncol=3, reduction = "umap")
```

## Save 

```{r}
saveRDS(seurat.rna, file="save/seurat_ko")
```

## 8. Integration 

### 8.0 Load 

```{r}
seurat_ctl <- readRDS("save/seurat_ctl")
seurat_denerv <- readRDS("save/seurat_denerv")
```

### 8.1 Merge

```{r}
seurat.merge <- merge(seurat_ctl, c(seurat_denerv)) %>%
  FindVariableFeatures(nfeatures = 3000) %>%
  ScaleData() %>%
  RunPCA(npcs = 50) %>%
  RunUMAP(dims = 1:20)
plot1 <- DimPlot(seurat.merge, group.by="orig.ident")
plot2 <- FeaturePlot(seurat.merge, c("Myh1","Myh2","Myh4","Myh7"), ncol=2, pt.size = 0.1)
plot1
plot2
```

### 8.2-A Using Seurat 

```{r}
seurat_objs <- list(DS1 = seurat_ctl, DS2 = seurat_denerv)
anchors <- FindIntegrationAnchors(object.list = seurat_objs, dims = 1:30)
seurat.integrate <- IntegrateData(anchors, dims = 1:30)
```

```{r}
seurat.integrate <- ScaleData(seurat.integrate)
seurat.integrate <- RunPCA(seurat.integrate, npcs = 50)
ElbowPlot(seurat.integrate, ndims = ncol(Embeddings(seurat.integrate, "pca")))
```

```{r}
seurat.integrate <- RunUMAP(seurat.integrate, dims = 1:15)
```

```{r}
DefaultAssay(seurat.integrate) <- "RNA"
plot1 <- UMAPPlot(seurat.integrate, group.by="orig.ident")
plot2 <- FeaturePlot(seurat.integrate, c("Myh1","Myh2","Myh4","Myh7"), ncol=2, pt.size = 0.1)

plot1
plot2
```

### 8.2-B Using Harmony 

```{r}
seurat.harm <- merge(seurat_ctl, c(seurat_ctk, seurat_denerv)) %>%
  FindVariableFeatures(nfeatures = 3000) %>%
  ScaleData() %>%
  RunPCA(npcs = 50)

seurat.harm <- RunHarmony(seurat.harm, group.by.vars = "orig.ident", dims.use = 1:20, max.iter.harmony = 50)
seurat.harm <- RunUMAP(seurat.harm, reduction = "harmony", dims = 1:20)
```

```{r}
DefaultAssay(seurat.harm) <- "RNA"
plot1 <- UMAPPlot(seurat.harm, group.by="orig.ident")
plot2 <- UMAPPlot(seurat.harm, label = T)
plot3 <- FeaturePlot(seurat.harm, c("Myh1","Myh2","Myh4","Myh7"), ncol=2, pt.size = 0.1)

plot1
plot2
plot3
```

### Save

```{r}
saveRDS(seurat.integrate, file="save/seurat_4integrate")
```

## 9. Clustering 

```{r}
seurat.rna <- seurat.integrate
seurat.rna$status <- NA 
seurat.rna$status[seurat.rna$orig.ident == "Ctl"] <- "WT"
# seurat.rna$status[seurat.rna$orig.ident == "Ctk"] <- "WT"
seurat.rna$status[seurat.rna$orig.ident == "Denerv"] <- "D"
# seurat.rna$status[seurat.rna$orig.ident == "KO"] <- "KO"
```

```{r}
DefaultAssay(seurat.rna) <- "integrated"
seurat.rna <- FindNeighbors(seurat.rna, dims = 1:ncol(Embeddings(seurat.rna)))
seurat.rna <- FindClusters(seurat.rna, resolution = 0.8)

DimPlot(seurat.rna, reduction = "umap", label = TRUE)
```

```{r}
DefaultAssay(seurat.rna) <- "integrated"
resolution.range <- seq(from = 0, to = 1.2, by = 0.2)
seurat.rna.clustree <- FindClusters(object = seurat.rna, resolution = resolution.range)
head(seurat.rna.clustree@meta.data)
clustree(seurat.rna.clustree, prefix = 'integrated_snn_res.')
```

## 10. Annotate cells 

```{r}
DefaultAssay(seurat.rna) <- "RNA"
input_data.markers <- FindAllMarkers(seurat.rna, min.pct = 0.25, logfc.threshold = 0.25, only.pos = TRUE)
input_data.markers %>% group_by(cluster) %>% slice_max(n = 5, order_by = avg_log2FC)
input_data.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC) -> topX
```

```{r}
plot1 <- FeaturePlot(seurat.rna, c("Myh1", "Myh2", "Myh4","Col22a1"), ncol = 2)
plot2 <- FeaturePlot(seurat.rna, c("Myh7", "Pam", "Uty","Gm28653"), ncol = 2)
plot3 <- VlnPlot(seurat.rna, features = c("Myh1", "Myh2", "Myh7","Col22a1"), pt.size = 0, ncol = 2)
plot1
plot2 
plot3
```

```{r}
new_ident <- setNames(c("myh1,cd36",
                        "smox,myh4",
                        "gm28653,smad4",
                        "ppara,myh1",
                        "dlg2,gm",
                        "xist,myhas",
                        "myh4,mybpc2",
                        "scn5a,cntfr",
                        "dlc1,abca8a",
                        "hectd2os,lmcd1",
                        "lama2, col22a1",
                        "ano4,prkar1a"),
                      levels(seurat.rna))
seurat.rna <- RenameIdents(seurat.rna, new_ident)
DimPlot(seurat.rna, reduction = "umap", label = TRUE) + NoLegend()
```


```{r}
seurat.rna$region <- factor(setNames(c("TypeMixt",
                                       "TypeIIb",
                                       "Denerv",
                                       "TypeIIa",
                                       "Denerv",
                                       "TypeIIb",
                                       "TypeIIb",
                                       "Denerv",
                                       "Other",
                                       "Denerv",
                                       "Other",
                                       "Other"),
                                 levels(seurat.rna@active.ident))[seurat.rna@active.ident],
                        levels=c("TypeIIa","TypeIIb", "TypeMixt", "Other", "Denerv"))
DimPlot(seurat.rna, reduction = "umap", group.by = "region") & NoAxes()
```

## 11. Composition

### 11.1 Visualize the compostions 

```{r}
freq <- table(seurat.rna$seurat_clusters, seurat.rna$status)
prop <- apply(freq,2,function(x) x/sum(x))
layout(matrix(1:3,nrow=1)); par(mar=c(8,5,1,1))
barplot(freq, col=c("#96bbb3","#ebd54c","#364437","#5e341e", "#992b1e"),
        border=NA, las=2, ylab="Frequency", cex.names = 0.8)
barplot(prop, col=c("#96bbb3","#ebd54c","#364437","#5e341e", "#992b1e"),
        border=NA, las=2, ylab="Proportion", cex.names = 0.8)
plot.new()
legend("left", fill=c("#96bbb3","#ebd54c","#364437","#5e341e", "#992b1e"), legend=rownames(freq), bty="n")
```

### 11.2-A Composition Comparaison using Fisher  


```{r}
freq_fisher <- function(conditions, have_identity){
  freq <- table(factor(have_identity, levels=c(TRUE,FALSE)),
              conditions)
  test <- fisher.test(freq)
  res <- setNames(c(test$estimate, test$p.value), c("oddsratio","pval_fisher"))
  return(res)
}

region_enrichment <- data.frame(region = levels(seurat.rna$region),
                                t(sapply(levels(seurat.rna$region), function(region)
                                  freq_fisher(conditions = factor(seurat.rna$status, levels=c("WT","D")),
                                              have_identity = seurat.rna$region == region)
                                  )),
                                row.names=NULL)
region_enrichment$padj_fisher <- p.adjust(region_enrichment$pval_fisher)
region_enrichment
```

### 11.2-B Composition Comparaison using GLM  

```{r}
freq_glm_aov <- function(samples, conditions, have_identity){
  sample_conditions <- unique(data.frame(sample = samples, condition = conditions))
  sample_conditions <- setNames(sample_conditions$condition, sample_conditions$sample)
  freq <- table(samples, factor(have_identity, levels=c(TRUE,FALSE)))
  m <- glm(freq ~ sample_conditions[rownames(freq)], family = "binomial")
  aov <- anova(m, test = "Chisq")
  res <- setNames(c(coef(m)[2], aov$Pr[2]), c("coef_glm","pval_aov"))
  return(res)
}

region_enrichment <- data.frame(region_enrichment,
                                t(sapply(levels(seurat.rna$region), function(region){
                                freq_glm_aov(samples = seurat.rna$orig.ident,
                                             conditions = factor(seurat.rna$status, levels=c("WT","D")),
                                             seurat.rna$region == region)
                                  })),
                                row.names=NULL)
region_enrichment$padj_aov <- p.adjust(region_enrichment$pval_aov)
region_enrichment
```

```{r}
org_status <- unique(data.frame(orig.ident=seurat.rna$orig.ident,
                                status=factor(seurat.rna$status, levels=c("WT","KO"))))
org_status <- setNames(org_status$status, org_status$orig.ident)

layout(matrix(1:6,nrow=1))

for(region in c("TypeIIa","TypeIIb", "TypeMixt", "Other", "Denerv")){
  props <- setNames(sapply(sort(unique(seurat.rna$orig.ident)), function(orig.ident)
    mean(seurat.rna$region[seurat.rna$orig.ident == orig.ident] == region)),
    sort(unique(seurat.rna$orig.ident)))
  barplot(props[order(props)],
          col = ifelse(org_status[names(props)[order(props)]] == "WT", "#cdcdcd", "#303030"),
          ylab = "Proportions", main = region, las = 2, cex.names = 0.8)
  }

plot.new()
legend("left", fill = c("#cdcdcd","#303030"), legend = c("WT","Dener"), bty="n")
```

##  12. Differential Expression 

### 12.1 Transcriptome similarity 

```{r}
seurat_0 <- subset(seurat.rna, subset = seurat_clusters == 0) %>%
  FindVariableFeatures()
avg_expr_org <- sapply(sort(unique(seurat.rna$orig.ident)), function(org)
  rowMeans(seurat_0@assays$RNA@data[,which(seurat_0$orig.ident == org)]))
corr_org <- cor(avg_expr_org[VariableFeatures(seurat_0),])
plot(hclust(as.dist(1-corr_org)))
```

### 12.2 Wilcoxon test for DE

```{r}
DE_wilcoxauc_0 <- wilcoxauc(seurat_0, group_by = "status") %>%
  filter(group == "D") %>%
  mutate(DE = abs(logFC)>log(1.1) & padj < 0.01) %>%
  mutate(DEG = ifelse(abs(logFC)>log(1.1) & padj < 0.01, feature, NA))
```

```{r}
ggplot(DE_wilcoxauc_0, aes(x = logFC, y = -log10(padj), col=DE, label=DEG)) +
  geom_point() +
  geom_text_repel() +
  geom_vline(xintercept=c(-log(1.1), log(1.1), 0), col="#303030", linetype="dotted") +
  geom_hline(yintercept=-log10(0.01), col="#303030", linetype="dotted") +
  scale_color_manual(values=c("#909090", "red")) +
  theme_minimal()
```

## 13. Pseudotemporal Analysis 

```{r}
seurat.type <- subset(seurat.rna, subset = RNA_snn_res.0.8 %in% c(1,2,4,5,6,7))
seurat.type <- FindVariableFeatures(seurat.type, nfeatures = 2000)
VariableFeatures(seurat.type) <- setdiff(VariableFeatures(seurat.type), unlist(cc.genes))
seurat.type <- CellCycleScoring(seurat.type,
                                s.features = cc.genes$s.genes,
                                g2m.features = cc.genes$g2m.genes,
                                set.ident = TRUE)
seurat.type <- ScaleData(seurat.type, vars.to.regress = c("S.Score", "G2M.Score"))
```

```{r}
seurat.type <- RunPCA(seurat.type) %>% RunUMAP(dims = 1:20)
FeaturePlot(seurat.type, c("Myh2", "Myh4", "Xist", "Uty"), ncol = 2)
```

```{r}
dm <- DiffusionMap(Embeddings(seurat.type, "pca")[,1:20])
dpt <- DPT(dm)
seurat.type$dpt <- rank(-dpt$dpt)
FeaturePlot(seurat.type, c("dpt", "Myh4", "Xist", "Uty"), ncol=2)
```

```{r}
plot1 <- qplot(seurat.type$dpt, as.numeric(seurat.type@assays$RNA@data["Myh4",]),
               xlab="Dpt", ylab="Expression", main="Myh4") +
  geom_smooth(se = FALSE, method = "loess") + theme_bw()
plot2 <- qplot(seurat.type$dpt, as.numeric(seurat.type@assays$RNA@data["Smox",]),
               xlab="Dpt", ylab="Expression", main="Smox") +
  geom_smooth(se = FALSE, method = "loess") + theme_bw()
plot3 <- qplot(seurat.type$dpt, as.numeric(seurat.type@assays$RNA@data["Cntfr",]),
               xlab="Dpt", ylab="Expression", main="Cntfr") +
  geom_smooth(se = FALSE, method = "loess") + theme_bw()
plot1
plot2
plot3
```

